File: CodexRemoteProbe.py
import os, json, time
from pathlib import Path
from datetime import datetime, timezone

home = Path(os.environ.get('PROBE_CODEX_HOME') or os.environ.get('CODEX_HOME') or (Path.home()/'.codex'))
max_sessions = max(1, int(os.environ.get('PROBE_MAX_SESSIONS','8')))
stale_minutes = max(1.0, float(os.environ.get('PROBE_STALE_MINUTES','30')))

def j(line):
    try: return json.loads(line)
    except Exception: return None

def index_map():
    m={}
    p=home/'session_index.jsonl'
    if not p.exists(): return m
    try:
        for line in p.read_text('utf-8', errors='ignore').splitlines():
            o=j(line)
            if not o: continue
            i=o.get('id') or o.get('session_id') or o.get('thread_id')
            t=o.get('thread_name') or o.get('title')
            if i and t: m[str(i)] = str(t)
    except Exception: pass
    return m

def first_json(p):
    try:
        with p.open('rb') as f:
            b=f.read(128*1024)
        return j(b.decode('utf-8','ignore').splitlines()[0])
    except Exception: return None

def scan_tail(p, max_bytes=32*1024*1024):
    life=None; token=None
    try:
        size=p.stat().st_size
        with p.open('rb') as f:
            pos=size; carry=b''; scanned=0
            while pos>0 and scanned<max_bytes and (life is None or token is None):
                n=min(256*1024,pos,max_bytes-scanned); pos-=n; scanned+=n
                f.seek(pos); data=f.read(n)+carry
                parts=data.splitlines()
                carry=parts[0] if parts else b''
                for raw in reversed(parts[1:] if len(parts)>1 else []):
                    s=raw.decode('utf-8','ignore')
                    if life is None and any(x in s for x in ('task_started','task_complete','turn_aborted')):
                        o=j(s); t=((o or {}).get('payload') or {}).get('type')
                        if (o or {}).get('type')=='event_msg' and t in ('task_started','task_complete','turn_aborted'):
                            life={'type':t,'timestamp':o.get('timestamp')}
                    if token is None and 'token_count' in s:
                        o=j(s)
                        if (o or {}).get('type')=='event_msg' and ((o or {}).get('payload') or {}).get('type')=='token_count':
                            token=o['payload']
                    if life is not None and token is not None: break
    except Exception: pass
    return life, token

def leaf(s):
    if not s: return ''
    return str(s).rstrip('/\\').replace('\\','/').split('/')[-1]

titles=index_map(); sessions=[]; latest_rl=None; latest_mtime=0
root=home/'sessions'
files=[]
if root.exists():
    try:
        for p in root.rglob('rollout-*.jsonl'):
            try: files.append((p,p.stat().st_mtime))
            except Exception: pass
    except Exception: pass
files.sort(key=lambda x:x[1], reverse=True)
for p,mtime in files:
    if len(sessions)>=max_sessions: break
    meta_line=first_json(p); meta=(meta_line or {}).get('payload') if (meta_line or {}).get('type')=='session_meta' else {}
    meta=meta or {}
    if meta.get('parent_thread_id') or meta.get('parentThreadId'): continue
    sid=str(meta.get('id') or meta.get('session_id') or meta.get('sessionId') or p.stem)
    life,token=scan_tail(p)
    age=(time.time()-mtime)/60.0; lt=(life or {}).get('type')
    status='unknown'
    if lt=='task_started': status='stale' if age>stale_minutes else 'running'
    elif lt=='task_complete': status='completed'
    elif lt=='turn_aborted': status='aborted'
    info=(token or {}).get('info') or {}; last=info.get('last_token_usage') or info.get('lastTokenUsage') or {}
    total=last.get('total_tokens', last.get('totalTokens')); ctx=info.get('model_context_window', info.get('modelContextWindow'))
    pct=round(min(100,float(total)/float(ctx)*100),1) if total is not None and ctx else None
    rl=(token or {}).get('rate_limits') or (token or {}).get('rateLimits')
    if rl and mtime>latest_mtime: latest_rl=rl; latest_mtime=mtime
    cwd=str(meta.get('cwd') or '')
    sessions.append({'id':sid,'title':titles.get(sid) or leaf(cwd) or ('Session '+sid[:8]),'workspace':leaf(cwd),'source':'WSL Docker','status':status,'lifecycle':lt,'updatedAt':datetime.fromtimestamp(mtime,timezone.utc).isoformat(),'contextPercent':pct,'totalTokens':total})
print(json.dumps({'ok':True,'codexHome':str(home),'processCount':None,'sessions':sessions,'latestRateLimits':latest_rl}, ensure_ascii=False))

File: CodexRemoteProbe.js
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const cp = require('child_process');

const codexHome = process.env.PROBE_CODEX_HOME || process.env.CODEX_HOME || path.join(os.homedir(), '.codex');
const maxSessions = Math.max(1, Number.parseInt(process.env.PROBE_MAX_SESSIONS || '8', 10) || 8);
const staleMinutes = Math.max(1, Number.parseFloat(process.env.PROBE_STALE_MINUTES || '30') || 30);
const maxScanBytes = 32 * 1024 * 1024;
const chunkSize = 256 * 1024;

function parseJson(line) {
  try { return JSON.parse(line); } catch { return null; }
}

function readIndex() {
  const map = new Map();
  const p = path.join(codexHome, 'session_index.jsonl');
  if (!fs.existsSync(p)) return map;
  const text = fs.readFileSync(p, 'utf8');
  for (const line of text.split(/\r?\n/)) {
    if (!line.trim()) continue;
    const obj = parseJson(line);
    if (!obj) continue;
    const id = obj.id || obj.session_id || obj.thread_id;
    const title = obj.thread_name || obj.title;
    if (id && title) map.set(String(id), String(title));
  }
  return map;
}

function listRollouts(root) {
  const out = [];
  if (!fs.existsSync(root)) return out;
  const stack = [{ dir: root, depth: 0 }];
  while (stack.length) {
    const { dir, depth } = stack.pop();
    let entries;
    try { entries = fs.readdirSync(dir, { withFileTypes: true }); } catch { continue; }
    for (const e of entries) {
      const p = path.join(dir, e.name);
      if (e.isDirectory() && depth < 5) stack.push({ dir: p, depth: depth + 1 });
      else if (e.isFile() && /^rollout-.*\.jsonl$/i.test(e.name)) {
        try {
          const st = fs.statSync(p);
          out.push({ path: p, mtimeMs: st.mtimeMs, size: st.size });
        } catch {}
      }
    }
  }
  out.sort((a, b) => b.mtimeMs - a.mtimeMs);
  return out;
}

function readFirstJsonLine(file) {
  const fd = fs.openSync(file, 'r');
  try {
    const buf = Buffer.alloc(128 * 1024);
    const n = fs.readSync(fd, buf, 0, buf.length, 0);
    const text = buf.subarray(0, n).toString('utf8');
    const first = text.split(/\r?\n/, 1)[0];
    return parseJson(first);
  } finally {
    fs.closeSync(fd);
  }
}

function scanTail(file, size) {
  const fd = fs.openSync(file, 'r');
  let pos = size;
  let scanned = 0;
  let carry = '';
  let lifecycle = null;
  let token = null;
  try {
    while (pos > 0 && scanned < maxScanBytes && (!lifecycle || !token)) {
      const len = Math.min(chunkSize, pos, maxScanBytes - scanned);
      pos -= len;
      scanned += len;
      const buf = Buffer.allocUnsafe(len);
      fs.readSync(fd, buf, 0, len, pos);
      const text = buf.toString('utf8') + carry;
      const lines = text.split(/\r?\n/);
      carry = lines.shift() || '';
      for (let i = lines.length - 1; i >= 0; i--) {
        const line = lines[i];
        if (!line) continue;
        if (!lifecycle && (line.includes('task_started') || line.includes('task_complete') || line.includes('turn_aborted'))) {
          const obj = parseJson(line);
          const t = obj && obj.type === 'event_msg' && obj.payload && obj.payload.type;
          if (t === 'task_started' || t === 'task_complete' || t === 'turn_aborted') {
            lifecycle = { type: t, timestamp: obj.timestamp || null, turnId: obj.payload.turn_id || null };
          }
        }
        if (!token && line.includes('token_count')) {
          const obj = parseJson(line);
          if (obj && obj.type === 'event_msg' && obj.payload && obj.payload.type === 'token_count') {
            token = obj.payload;
          }
        }
        if (lifecycle && token) break;
      }
    }
    if ((!lifecycle || !token) && carry) {
      const obj = parseJson(carry);
      if (obj && obj.type === 'event_msg' && obj.payload) {
        const t = obj.payload.type;
        if (!lifecycle && (t === 'task_started' || t === 'task_complete' || t === 'turn_aborted')) {
          lifecycle = { type: t, timestamp: obj.timestamp || null, turnId: obj.payload.turn_id || null };
        }
        if (!token && t === 'token_count') token = obj.payload;
      }
    }
  } finally {
    fs.closeSync(fd);
  }
  return { lifecycle, token };
}

function leaf(p) {
  if (!p) return '';
  const clean = String(p).replace(/[\\/]+$/, '');
  const parts = clean.split(/[\\/]/);
  return parts[parts.length - 1] || clean;
}

function processCount() {
  try {
    const txt = cp.execSync('ps -eo pid=,args=', { encoding: 'utf8', stdio: ['ignore', 'pipe', 'ignore'] });
    return txt.split(/\r?\n/).filter(line => {
      const s = line.toLowerCase();
      return s.includes('codex') && !s.includes('codexremoteprobe') && !s.includes('grep codex');
    }).length;
  } catch { return null; }
}

const titles = readIndex();
const files = listRollouts(path.join(codexHome, 'sessions'));
const sessions = [];
let newestRateLimits = null;
let newestRateLimitsMtime = 0;

for (const f of files) {
  if (sessions.length >= maxSessions) break;
  let metaLine;
  try { metaLine = readFirstJsonLine(f.path); } catch { continue; }
  const meta = metaLine && metaLine.type === 'session_meta' ? (metaLine.payload || {}) : {};
  if (meta.parent_thread_id || meta.parentThreadId) continue;
  const id = String(meta.id || meta.session_id || meta.sessionId || path.basename(f.path).replace(/^rollout-.*-([0-9a-f-]{16,})\.jsonl$/i, '$1'));
  let scan;
  try { scan = scanTail(f.path, f.size); } catch { scan = { lifecycle: null, token: null }; }
  const ageMinutes = (Date.now() - f.mtimeMs) / 60000;
  let status = 'unknown';
  const life = scan.lifecycle && scan.lifecycle.type;
  if (life === 'task_started') status = ageMinutes > staleMinutes ? 'stale' : 'running';
  else if (life === 'task_complete') status = 'completed';
  else if (life === 'turn_aborted') status = 'aborted';

  const info = scan.token && scan.token.info;
  const lastUsage = info && info.last_token_usage;
  const totalTokens = lastUsage && (lastUsage.total_tokens ?? lastUsage.totalTokens);
  const contextWindow = info && (info.model_context_window ?? info.modelContextWindow);
  const contextPercent = (Number.isFinite(Number(totalTokens)) && Number(contextWindow) > 0)
    ? Math.min(100, Math.round((Number(totalTokens) / Number(contextWindow)) * 1000) / 10)
    : null;

  const rateLimits = scan.token && scan.token.rate_limits;
  if (rateLimits && f.mtimeMs > newestRateLimitsMtime) {
    newestRateLimits = rateLimits;
    newestRateLimitsMtime = f.mtimeMs;
  }

  const cwd = meta.cwd ? String(meta.cwd) : '';
  sessions.push({
    id,
    title: titles.get(id) || leaf(cwd) || `Session ${id.slice(0, 8)}`,
    workspace: leaf(cwd),
    source: 'WSL Docker',
    originator: meta.originator || null,
    status,
    lifecycle: life || null,
    updatedAt: new Date(f.mtimeMs).toISOString(),
    contextPercent,
    totalTokens: totalTokens == null ? null : Number(totalTokens)
  });
}

process.stdout.write(JSON.stringify({
  ok: true,
  codexHome,
  processCount: processCount(),
  sessions,
  latestRateLimits: newestRateLimits
}));
